### PR TITLE
[15477] Update BasicConfigurationExample to allow set up TTL

### PR DESCRIPTION
**Pull request by MiguelBarro**
## Description

Command line processing now relies entirely on *optionparser.h* capabilities.
A new `--ttl` option has been added that specifies how many hops must multicast specify. It doesn't affect unicast TTL.
If no transport is specified but ttl is then the example will try to mimick builtin transport with the specified hops.

## Contributor Checklist
- [x] Commit messages follow the project guidelines. 
- [x] The code follows the style guidelines of this project. 
- [] NA Tests that thoroughly check the new feature have been added/Regression tests checking the bug and its fix have been added. 
- [x] Any new/modified methods have been properly documented using Doxygen. 
- [x] Fast DDS test suite has been run locally. 
- [x] Changes are ABI compatible. 
- [x] Changes are API compatible. 
- [x] Documentation builds and tests pass locally. 
- [ ] NA New feature has been added to the `versions.md` file (if applicable).
- [x] New feature has been documented/Current behavior is correctly described in the documentation. 

## Reviewer Checklist
- [x] Check contributor checklist is correct.
- [x] Check CI results: changes do not issue any warning.
- [x] Check CI results: failing tests are unrelated with the changes.

### CODE REVIEW

File: examples/cpp/dds/BasicConfigurationExample/BasicConfiguration_main.cpp [hunk 1]
```diff
@@ -60,154 +60,155 @@ int main(
...
+            case optionIndex::RELIABLE:
+                reliable = true;
+                break;
+
+            case optionIndex::TRANSIENT_LOCAL:
+                transient = true;
+                break;
+
+            case optionIndex::TTL:
+                hops = strtol(opt.arg, nullptr, 10);
```
Comment (juanlofer-eprosima) [inline examples/cpp/dds/BasicConfigurationExample/BasicConfiguration_main.cpp:190]: Maybe add a message warning that this argument will take no effect when transport is `SHM`.

Also, it would be good to have a warning somewhere, or in the help menu, informing about the `255` hard limit.

File: examples/cpp/dds/BasicConfigurationExample/BasicConfigurationPublisher.cpp [hunk 2]
```diff
@@ -78,24 +79,53 @@ bool HelloWorldPublisher::init(
...
+            case UDPv6:
+            {
+                auto udp_transport = std::make_shared<UDPv6TransportDescriptor>();
+                pqos.transport().user_transports.push_back(udp_transport);
+            }
+            break;
+            case DEFAULT:
+            default:
+            {
+                // mimick default transport selection
```
Comment (juanlofer-eprosima) [inline examples/cpp/dds/BasicConfigurationExample/BasicConfigurationPublisher.cpp:109]: Isn't it possible to directly modify the builtin transports?

**Review by juanlofer-eprosima — CHANGES_REQUESTED** [review]

File: examples/cpp/dds/BasicConfigurationExample/BasicConfigurationPublisher.cpp (hunk 2 above)
Comment (MiguelBarro) [reply examples/cpp/dds/BasicConfigurationExample/BasicConfigurationPublisher.cpp:109]: nope

File: examples/cpp/dds/BasicConfigurationExample/BasicConfiguration_main.cpp [hunk 3]
```diff
@@ -60,154 +60,155 @@ int main(
...
+    }
+
+    if (options[optionIndex::HELP])
+    {
+        option::printUsage(fwrite, stdout, usage, columns);
+        return 0;
+    }
+
+    // Decide between publisher or subscriber
+    if (parse.nonOptionsCount())
```
[the remark this answers is not in the document]
Comment (MiguelBarro) [reply examples/cpp/dds/BasicConfigurationExample/BasicConfiguration_main.cpp]: fixed

File: examples/cpp/dds/BasicConfigurationExample/BasicConfiguration_main.cpp (hunk 1 above)
Comment (MiguelBarro) [reply examples/cpp/dds/BasicConfigurationExample/BasicConfiguration_main.cpp:200]: fixed

File: examples/cpp/dds/BasicConfigurationExample/BasicConfiguration_main.cpp [hunk 4]
```diff
@@ -60,145 +60,166 @@ int main(
...
+    }
+
+    if (transport == SHM && hops > 0 )
+    {
+        std::cerr << "WARNING: --ttl will take no effect since not using UDP transport." << std::endl;
+    }
+
+    if (hops > 255 )
+    {
+        std::cerr << "WARNING: --ttl value will be cut off to it's maximum 255." << std::endl;
```
Comment (juanlofer-eprosima) [inline examples/cpp/dds/BasicConfigurationExample/BasicConfiguration_main.cpp:222]:
```suggestion
        std::cerr << "WARNING: --ttl value will be cut off to its maximum 255." << std::endl;
```

File: examples/cpp/dds/BasicConfigurationExample/BasicConfiguration_main.cpp [hunk 5]
```diff
@@ -60,145 +60,166 @@ int main(
...
+
+    // Decide between publisher or subscriber
+    try
+    {
+        if (parse.nonOptionsCount() != 1)
         {
-            std::cerr << "ERROR: first argument can only be <publisher|subscriber>" << std::endl;
-            option::printUsage(fwrite, stdout, usage, columns);
-            return 1;
+            throw 1;
```
Comment (juanlofer-eprosima) [inline examples/cpp/dds/BasicConfigurationExample/BasicConfiguration_main.cpp:91]: NIT: print a different trace for this error?

**Review by juanlofer-eprosima — CHANGES_REQUESTED** [review]

File: examples/cpp/dds/BasicConfigurationExample/BasicConfiguration_main.cpp (hunk 4 above)
Comment (MiguelBarro) [reply examples/cpp/dds/BasicConfigurationExample/BasicConfiguration_main.cpp:222]: done

File: examples/cpp/dds/BasicConfigurationExample/BasicConfiguration_main.cpp (hunk 5 above)
Comment (MiguelBarro) [reply examples/cpp/dds/BasicConfigurationExample/BasicConfiguration_main.cpp:91]: or maybe we can rephrase it to be valid on all cases

**Review by juanlofer-eprosima — APPROVED** [review]
LGTM

**Comment by MiguelCompany** [comment]
@richiprosima Please test this
